In [1]:
import pandas as pd
import json

In [2]:
reviews_json_file_path = 'Pet_Supplies.json'
df = pd.read_json(reviews_json_file_path, lines=True)

print("Number of rows in the dataset:", df.shape[0])
print("Number of columns in the dataset:", df.shape[1])

Number of rows in the dataset: 6542483
Number of columns in the dataset: 12


In [9]:
df.head()

overall vote  verified   reviewTime      reviewerID        asin  \
0        3    5      True  09 27, 2007  A13K4OZKAAHOXS  0972585419   
1        4    6      True  08 25, 2007  A1DWYEX4P7GB7Z  0972585419   
2        4    5      True  02 20, 2007  A3NVN97YJSKEPC  0972585419   
3        1  NaN      True  12 30, 2016  A1PDMES1LYA0DP  0972585419   
4        1  NaN      True  12 23, 2016   AT6BH0TQLZS5X  0972585419   

       reviewerName                                         reviewText  \
0    100indecisions  Either my bird can't talk, or it really doesnt...   
1         J. Weaver  The CD is a good idea for teaching a bird to s...   
2  Theresa Ehrhardt  good cd to teach birds to talk and learn new w...   
3             Kathi  I was not happy with product would like to ret...   
4            Alyssa  This cd is scratched and it constantly skips. ...   

                           summary  unixReviewTime style image  
0  Still waiting for it to work...      1190851200   NaN   NaN  
1       Feathered Phonics volume 1      1188000000   NaN   NaN  
2                          bird cd      1171929600   NaN   NaN  
3                         One Star      1483056000   NaN   NaN  
4                         One Star      1482451200   NaN   NaN

#### The first 100 users with the most purchases with the corresponding number of purchases (NumberOfReviews column) and the number of users who purchased the same products (NumberOfUsers column)

In [ ]:
reviewer_counts = df['reviewerID'].value_counts()
reviewer_counts_df = reviewer_counts.reset_index()

In [10]:
pd.set_option('display.max_rows', None)

In [11]:
top_reviewers = reviewer_counts_df.head(100)['reviewerID']
top_reviewer_ids = top_reviewers.tolist()
results = []

for reviewer_id in top_reviewer_ids:
    reviewer_df = df[df['reviewerID'] == reviewer_id]
    num_reviews = len(reviewer_df)

    unique_asin_values = reviewer_df['asin'].unique()
    filtered_df = df[df['asin'].isin(unique_asin_values)]
    num_reviewer_ids = filtered_df['reviewerID'].nunique()
    
    results.append({
        'reviewerID': reviewer_id,
        'NumberOfReviews': num_reviews,
        'NumberOfUsers': num_reviewer_ids
    })

results_df = pd.DataFrame(results)
results_df

reviewerID  NumberOfReviews  NumberOfUsers
0   A11OTLEDSW8ZXD              504         176294
1    A8K5I4MF6QCGJ              310          20031
2   A22CW0ZHY3NJH8              294         148105
3   A3ICUM9JY3HLG3              293         128673
4   A2GJX2KCUSR0EI              288         130792
5   A2O421DTA8J0RW              241         118084
6   A2M9GBUPEVJ5F7              229         132906
7   A175XXVMLQA98A              221          47233
8   A11FX8HL2ANK6T              214         101131
9   A2A9X58G2GTBLP              211          36075
10  A2YKWYC3WQJX5J              204          69753
11  A2503LT8PZIHAD              204          87927
12  A100WO06OQR8BQ              189         111449
13  A21U4DR8M6I9QN              189         135173
14  A35R32TA60XD57              185          62651
15  A2LWPTMTWPWH2B              176          30203
16  A1DH3QDZX8Z7GO              176         121257
17  A3TAS1AG6FMBQW              175          98918
18  A2DQQUEPK53ORW              169          92059
19  A25QBCHO0KFT0P              167          75497
20   A5Z6TIROUVA5O              160          77445
21  A2FCGDCV199Q7S              157          26228
22  A2C9XE9I8RSKNX              156          95628
23  A3MXDIJVL9M9GQ              155          59187
24  A29WZU106L5RI8              154          64865
25   AYQ14XGJODM9J              154          97322
26  A1QBOC76MIOJYP              153          57133
27  A2RYWPOL4NN2KG              153          50213
28   AKBXK85W9S71P              149          81002
29   A8INT3NMHRUS4              147          46713
30  A2V1J3JT5OOZFO              147          89982
31  A2OCDK0BOW6UCY              147          31981
32   AQ8IGZ0FADU02              146          20236
33   AF8VTQ4MKAUZI              145          65213
34  A1Z4LX3Q1YMF84              144          46394
35  A1JZFGZEZVWQPY              140          66498
36  A288D2R9WAJBZS              139         102601
37  A1PKYIWTEQ1INI              139          64268
38   AT2JC2U5E474D              139          46167
39  A2BSKT65F16LAQ              138          68559
40  A1VHIQMUY0WYB2              137          63763
41   A3U2NAO7853MW              136         121961
42  A2V3THN58CDZLC              132          95060
43  A1PI8VBCXXSGC7              131          49994
44  A1RCG9K9AIPC3C              131          38807
45  A2RDLSG5LGD12E              130          63899
46  A1BV7JS2VQUANN              130         108057
47  A3Q9WQ73FWUVFT              130         110162
48   A5V34G66JTI66              129          63538
49  A2BH04B9G9LOYA              128         117588
50  A26W1UX7E3AW3J              126          31040
51  A1ZKGSDVP057L9              126          54549
52  A1RT0UQRCB2765              126          61314
53  A1XR2K3KNHMC8A              126          44692
54   APW4K3413QGI0              125         104660
55  A1F7YU6O5RU432              125          32159
56  A28Q0JA9H8WZOD              125          97776
57  A1F30VO5LDKGKG              125          93076
58  A13H5US9Z7L9M7              124          83094
59  A21B8AV7E3MPXE              124          32868
60  A12BTR2MVK2BR5              123          54405
61  A28EMTHVF120XV              121          48543
62   AMYTL79JMGQ6D              121          40662
63   AVZQ08UZU1YHA              121          74628
64  A2LOPO78PWU6NN              121          74615
65   ARPL5FW4OQ1YN              120         104379
66   ATJN4KEHJBOC8              120          40893
67   ABQVDKWU40G0V              120          38131
68   A8CGWGJ9OM58J              119          63625
69  A2UO040HWOP0C2              118          77851
70  A1F2T6RK47WKNC              116          79542
71  A2RUQOHBT1X3X6              115          35160
72    AQZIV618AOCU              115         115364
73  A1VOIQ29Q1EO1V              114          66885
74   AC5GLGVJXDEFV              114          81228
75  A2IV6TCBFJQRQN              114         103462
76  A2DFM26VLNVYNY              114         116109
77  A1ZPY91VE3IDN1            

In [12]:
pd.set_option('display.max_rows', 10)

#### We have selected the following 15 users
- **A11OTLEDSW8ZXD** (504, 176294) (users-1.json)
- **A22CW0ZHY3NJH8** (294, 148105) (users-2.json)
- **A21U4DR8M6I9QN** (189, 135173) (users-3.json)
- **A3U2NAO7853MW** (136, 121961) (users-4.json)
- **A100WO06OQR8BQ** (189, 111449) (users-5.json)
- **A288D2R9WAJBZS** (139, 102601) (users-6.json)
- **A3CG93783LP0FO** (110, 90534) (users-7.json)
- **AC5GLGVJXDEFV** (114, 81228) (users-8.json)
- **AQYKV5QEAO7P5** (111, 71756) (users-9.json)
- **A1RT0UQRCB2765** (126, 61314) (users-10.json)
- **A2RYWPOL4NN2KG** (153, 50213) (users-11.json)
- **A9NXTC0HTPSDO** (107, 40966) (users-12.json)
- **A26W1UX7E3AW3J** (126, 31040) (users-13.json)
- **A1HE5FWLYHKPQW** (108, 19728) (users-14.json)
- **A8K5I4MF6QCGJ** (310, 20031) (users-15.json)

#### Filtering the purchases of only a specific user

In [13]:
reviewerID = "A22CW0ZHY3NJH8"
specific_reviewer_df = df[df['reviewerID'] == reviewerID]
specific_reviewer_df.head()

overall vote  verified   reviewTime      reviewerID        asin  \
12937        4  NaN     False  11 12, 2010  A22CW0ZHY3NJH8  B00006H36X   
35330        3  NaN     False  04 20, 2016  A22CW0ZHY3NJH8  B000084DWM   
54999        3  NaN      True   05 9, 2010  A22CW0ZHY3NJH8  B000084EEF   
66964        4    3     False  08 27, 2012  A22CW0ZHY3NJH8  B00008URR8   
67537        5  NaN     False  08 20, 2017  A22CW0ZHY3NJH8  B00009NCZP   

      reviewerName                                         reviewText  \
12937           ()  I'm so happy to see I could buy Advantage thro...   
35330           ()  Truth be told, my cat is not finicky.  She get...   
54999           ()  I couldn't resist buying the Turbo Scratcher a...   
66964           ()  <div id="video-block-R2Q37Z2QSWG0VR" class="a-...   
67537           ()  Yum, yum, yum.  My cat loves canned food.  It'...   

                                              summary  unixReviewTime  \
12937                     My preferred flea treatment      1289520000   
35330                               UPDATE: Hairballs      1461110400   
54999                      Maybe good for other cats.      1273363200   
66964                              A true SPOT lifter      1346025600   
67537  It's even better when the meat comes in chunks      1503187200   

                                                   style image  
12937  {'Size:': ' 4 applications', 'Color:': ' over ...   NaN  
35330  {'Size:': ' 7 lb', 'Style:': ' Multi-Cat Dry |...   NaN  
54999                                                NaN   NaN  
66964                {'Style Name:': ' Spotlifter Only'}   NaN  
67537  {'Size:': ' 5.5 oz, 12 Pack', 'Flavor Name:': ...   NaN

#### The number of unique products purchased by this single specific

In [14]:
unique_asin_count = specific_reviewer_df['asin'].nunique()
print("Number of unique products:", unique_asin_count)

Number of unique products: 277


#### The number of users who purchased the same products

In [15]:
unique_asin_values = specific_reviewer_df['asin'].unique()
filtered_df = df[df['asin'].isin(unique_asin_values)]
reviewer_ids = filtered_df['reviewerID'].unique()
reviewer_ids_list = reviewer_ids.tolist()
print("Number of users who purchased the same products:", len(reviewer_ids))

Number of users who purchased the same products: 148105


#### Iteration through all identified users and for each product that was purchased by both an identified user and our specific user (one of 15 selected), we searched for 2 products (chronologically ordered) that the identified user had purchased before buying the given product.

In [ ]:
result_list = []

for reviewer_id in reviewer_ids_list:
    reviewer_df = df[df['reviewerID'] == reviewer_id]

    for product_id in unique_asin_values:
        sorted_products = reviewer_df.sort_values(by='unixReviewTime').reset_index(drop=True)
        product_indices = sorted_products[sorted_products['asin'] == product_id].index

        if not product_indices.empty and product_indices.min() >= 2:
            selected_products = sorted_products.loc[product_indices.min() - 2:product_indices.min() - 1, 'asin'].tolist()
            result_list.append({'reviewerID': reviewer_id, 'productID': product_id, 'selected_products': selected_products})

In [ ]:
with open('users-2.json', 'w') as json_file:
    json.dump(result_list, json_file)

#### This is how we collected data (sequences of length 2 and targets) for each of the selected 15 users on individual 7 instances of Google cloud computing service with pre-installed JupyterLab 3 environment.